New destroy handler

ReyhaneAskari opened the pull request:
Issue #5976. Supervisor class is kept but we added a has_destroyers attribute to fgraph, but some tests regarding dumping the pickle failed because a new attribute was added to the fgraph. I changed it like this : 
```python
fgraph.destroyers = [get_destroyers_of, has_destroyers]
```

@nouiz  what do you suggest?

fix https://github.com/Theano/Theano/issues/6230

lamblin commented:
Now that #5794 has been merged, you will probably need to rebase on the master. That will make the diff easier to check.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,6 +394,17 @@ def get_destroyers_of(r):
+            for item in l:
+                droot, _, root_destroyer = self.refresh_droot_impact()
+                try:
+                    [root_destroyer[droot[item]]]
```
I think you wanted this line to be:

```
if root_destroyer[droot[item]]:
     return True
```

ReyhaneAskari replied:
yeah! thanks.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,6 +394,17 @@ def get_destroyers_of(r):
+                try:
+                    [root_destroyer[droot[item]]]
+                    return True
+                except Exception:
```
I know you copied the Exception from above, but this is a bad practice. We should try to catch a much more precise exception.

Anyway, this isn't the optimized algorythm anyway. Both otherwise, you seem to have done the first part well. If with my modif in the other comment and tests pass, the structural change would be done. It would only miss the optimized implementation here.

ReyhaneAskari replied:
You are right! Thanks I changed it to ```KeyError```

ReyhaneAskari commented on theano/compile/function_module.py:
```diff
@@ -135,10 +135,15 @@ def __init__(self, protected):
-        for r in self.protected + list(fgraph.outputs):
-            if fgraph.destroyers(r):
+        if config.cycle_detection == 'fast':
+            if fgraph.destroyers[1](self.protected + list(fgraph.outputs)):
```
@nouiz , do we need to pass ```fgraph.outputs``` here ?

nouiz replied:
I think so.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,21 +394,25 @@ def get_destroyers_of(r):
+            clients_set = set()
             for protected_var in protected_list:
-                if recursive_destroys_finder(protected_var.clients):
+                clients_set.update(protected_var.clients)
```
the clients list of tuple (app, idx) is already a set.

nouiz replied:
So mostly, that commit don't help. There is some saving possible if one node reuse the same variable multiple time as input, but it happen so infrequently that I don't think the extra complexity is useful.

ReyhaneAskari commented:
I removed that commit. Now we need to check the failing tests and make sure they are the same tests that are failing in the master with ```THEANO_FLAGS=cycle_detetion='fast'```.  @lamblin suggested that I mark the failing tests in master as nosetests' known failures.

nouiz commented:
Just to be sure, make them as known failure only when cycle='fast'

On Mon, Jun 5, 2017 at 1:25 PM Reyhane Askari <notifications@github.com>
wrote:

> I removed that commit. Now we need to check the failing tests and make
> sure they are the same tests that are failing in the master with
> THEANO_FLAGS=cycle_detetion='fast'. @lamblin <https://github.com/lamblin>
> suggested that I mark the failing tests in master as nosetests' known
> failures.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6000#issuecomment-306250234>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC--UcopxR5lhHFzJjFo5wzs1_ggz3ks5sBDoLgaJpZM4Ntcs->
> .
>

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,6 +394,31 @@ def get_destroyers_of(r):
+        def has_destroyers(protected_list):
+            visited_app_set = set()
+
+            def recursive_destroys_finder(clients_list):
```
I would change the interface to have this function take a variable as input. Then it would loop over var.clients itself. I think it is a nicer interface and will help the conversion to the recursion free version.

ReyhaneAskari replied:
Sure. Thanks.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,6 +394,31 @@ def get_destroyers_of(r):
+                    destroy_maps = getattr(app.op, 'destroy_map', {}).values()
+                    if idx in [dmap for sublist in destroy_maps for dmap in sublist]:
+                        return True
+                    for var in getattr(app.op, 'view_map', {}).keys():
```
Here, this is too much restrictif. We are asking that all view of app are not destroyed. But we only care the outputs that are a view of app.inputs[idx].

ReyhaneAskari replied:
I see. Thanks. Is this check good? 
``` python
for var in getattr(app.op, 'view_map', {}).keys():
    if idx in app.op.view_map[var] and recursive_destroys_finder(app.outputs[var]):
          return True

```

nouiz replied:
that seem good. To help read it, I would rename var to var_idx, to tell it is an index, not a variable.

ReyhaneAskari replied:
Sure. Thanks.

ReyhaneAskari commented:
@lamblin, @nouiz I could only find ```unittest.expectedFailure``` as a decorator. To make sure it only adds when ```cycle_detection='fast'```, I wrote another decorator like this because there were no builtin options:

```python
import unittest
from theano import config

def expectedFailure_fast():
    return unittest.expectedFailure if config.cycle_detection == 'fast' else lambda x: x

@expectedFailure_fast()
def test_usage_loop_through_views_2():
```
Do you have any suggestions on where to add this code snippet? I was thinking maybe ```theano.gof.utils```?

nouiz commented:
theano.gof.utils seem the right place for that.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -409,12 +409,12 @@ def recursive_destroys_finder(clients_list):
-                        if recursive_destroys_finder(app.outputs[var].clients):
-                            return True
+                        if idx in app.op.view_map[var] and recursive_destroys_finder(app.outputs[var]):
+                                return True
```
you added an extra indentation that isn't useful.

nouiz commented:
theano.tests.unittests_tools would be better then theano.gof.utils I think.

lamblin commented:
Most issues seem to be:
```
 Error Details

OutputGuard.0 is a view/destroyed version of more then one inputs. Currently, we only support the case where an output is a view or a destroyed version of one input.

 Stack Trace

Traceback (most recent call last):
  File "/miniconda/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/jenkins/workspace/Theano_PR/theano/typed_list/tests/test_basic.py", line 257, in test_inplace
    accept_inplace=True)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/function.py", line 326, in function
    output_keys=output_keys)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/pfunc.py", line 486, in pfunc
    output_keys=output_keys)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/function_module.py", line 1814, in orig_function
    output_keys=output_keys)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/function_module.py", line 1491, in __init__
    insert_deepcopy(fgraph, inputs, outputs + additional_outputs)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/function_module.py", line 1099, in insert_deepcopy
    view_tree_set(alias_root(fgraph.outputs[i]), views_of_output_i)
  File "/home/jenkins/workspace/Theano_PR/theano/compile/function_module.py", line 57, in alias_root
    str(v) + " is a view/destroyed version of more then one inputs. "
NotImplementedError: OutputGuard.0 is a view/destroyed version of more then one inputs. Currently, we only support the case where an output is a view or a destroyed version of one input.
```

nouiz commented on theano/tensor/tests/test_sharedvar.py:
```diff
@@ -27,6 +27,7 @@ def makeSharedTester(shared_constructor_,
                      theano_fct_,
                      ref_fct_,
                      cast_value_=np.asarray,
+                     need_decorator=True,
```
Don't forget @lamblin comment in gh-6014.

ReyhaneAskari replied:
Yes, I'm on it.

ReyhaneAskari commented:
There was a small conflict here with jenkins and travis wouldn't run with out resolving it. I resolved the conflict online committed it. The name was automatically generated as ```Merge branch 'master' into new_destroy_handler```.  I will have to rebase in the end and I will clean this up.

ReyhaneAskari commented:
@lamblin @nouiz 
There were so many tests that were failing due to the insertion of ```OutputGurad```. I checked several tests and I figured that we should raise an error when [AdddestoryHandler tries to insert the ```OutputGurad``` ](https://github.com/Theano/Theano/blob/be022807cc9e3f8115a01ad2343c57f819f13ad9/theano/compile/mode.py#L116). So we need to check the ```graph.outputs``` as well as the ```protected_var```.  I also changed the ```has_destroyer``` such that in the case when the ```protected_var``` is an outputGuard, we do the check. 

I was not sure if there are cases where ```app == 'output'``` but the ```protected_var``` is not an OutputGuard. There are still some failing tests that I'm looking into.

ReyhaneAskari commented:
There were many failures in the tests because we were inplace in some cases where we shouldn't have been. I checked these tests and found out that we are not raising the error in the ```Supervisor``` class in some cases that the inplace was wrong. The problem lied in the case where we had a graph in which one node was cached in the ```visited_app_set``` and never checked again. 

The mechanism of ```visited_app_set``` is wrong. Consider the following case when an apply node is the client of serval variable nodes and has a destroy map of {0 : [1]}. In the first visit to this node since the destroy map is on the second input, we do not raise any error and we mark the apply node as visited while it should be checked for the second variable node as well where it is destroying one of its inputs. 

![photo_2017-06-12_13-55-52](https://user-images.githubusercontent.com/11895018/27047741-22fbf55a-4f77-11e7-8e68-596e43b44e64.jpg)

I guess it will be too complicated to support these cases. For now I removed the ```visited_app_set``` to see if I all the tests pass. Then in the benchmarking we can figure where is the bottleneck.

lamblin commented:
Conclusion of IRL discussion:
- remove view_map from OutputGuard
- In AddDestroyHandler, do not try to introduce OutputGuard when fast cycle detection is on
- undo the new check (full=True).

ReyhaneAskari commented:
We won't remove the view_map for the output Guard for now since it's causing many errors. We will handle that in another PR where we'll completely remove the output Guard.

ReyhaneAskari commented:
Is there a way to add the flag to the examples inside the docs? Maybe the easier way to change the cycle_detection flag was to change the default value to ```fast``` for the testing purposes and revert it back in the end.

nouiz commented:
Doing your suggestion would have done it.

On Tue, Jun 13, 2017 at 2:17 PM Reyhane Askari <notifications@github.com>
wrote:

> Is there a way to add the flag to the examples inside the docs? Maybe the
> easier way to change the cycle_detection flag was to change the default
> value to fast for the testing purposes and revert it back in the end.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6000#issuecomment-308203473>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-68x2ky6NgOogQxIMwQgvETlSxufks5sDtIngaJpZM4Ntcs->
> .
>

ReyhaneAskari commented:
I cleaned up some commits and also removed the changes to jenkins and travis and changed the cycle_detection flag to have ```fast``` as the default value.

ReyhaneAskari commented:
Just to make sure we don't have any errors on the master.

ReyhaneAskari commented:
Here is the results of profiling:

|               |   Function.call    | total compile time | Opt time    | Linker time   | Time since theano import |
|---------------|--------------------|--------------------|-------------|---------------|--------------------------|
| [master_2](https://github.com/Theano/Theano/files/1076079/master_2.txt)                  | 2317.635           | 586.9048           | 553.4296      | 22.69159                 | 3075.815      |
| [master_3](https://github.com/Theano/Theano/files/1076080/master_3.txt)                  | 2320.0             | 578.2175           | 545.9906      | 21.54995                 | 3058.587      |
| **avg master_2 master_3** | **2318.8175**      | **582.56115**      | **549.7101**  | **22.12077**             | **3067.201**  |
| [faster_2](https://github.com/Theano/Theano/files/1076078/faster_2.txt)                  | 2332.769           | 300.8603           | 269.1528      | 20.46626                 | 2796.888      |
| [faster_3](https://github.com/Theano/Theano/files/1076077/faster_3.txt)                  | 2334.987           | 298.6716           | 266.9097      | 20.59228                 | 2802.241      |
| **avg faster_2 faster_3** | **2333.878**       | **299.76595**      | **268.03125** | **20.52927**             | **2799.5645** |

Memory profiling:

||master|faster|
|----|--|--|
| peak memory cpu                      |       3.0 |       2.0 |
| peak memory gpu                      | 2296853.0 | 2186307.0 |
| peak memory cpu + gpu                | 2296855.0 | 2186309.0 |
| flag = optimizer_excluding cpu       |       3.0 |       2.0 |
| flag = optimizer_excluding gpu       | 2313898.0 | 2199204.0 |
| flag = optimizer_excluding cpu + gpu | 2313901.0 | 2199206.0 |
| allow_gc=False cpu                   |      47.0 |      47.0 |
| allow_gc=False gpu                   | 8456750.0 | 8693082.0 |
| allow_gc=False cpu + gpu             | 8456797.0 | 8693129.0 |

I also did the python profiling you can check them in the attached files. ```get_destroyers_of``` has the largest cumulative time in faster. Can we remove all the calls to it?

nouiz commented on theano/tensor/opt.py:
```diff
@@ -265,8 +265,8 @@ def apply(self, fgraph):
                                     not isinstance(node.inputs[i], Constant) and
-                                    # Is next line costly?
-                                    not fgraph.destroyers(node.inputs[i]) and
+                                    # Is next line costly? (used to be fgraph.get_destroyers)
```
I would update the comment to:

the next line should not be costly most of the time.

ReyhaneAskari replied:
Sure. Thanks.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -394,6 +394,32 @@ def get_destroyers_of(r):
                 return []
         fgraph.destroyers = get_destroyers_of
 
+        def has_destroyers(protected_list):
```
I think I would do something like this at the start of this method:

```
if not self.stale_droot:
    for var in protected_var:
        if self.destroyers(var):
            raise ...
```

This do 2 things:
1) We should always call has_destroyers what ever the value of cycle_detection is.
2) This could speed up or slow down the case when cycle_detection is the default value. So we need to speed test that. If is don't speed up computation, we should change the if to be `if self.cycle_detection != 'fast':`

What do you think of that?

ReyhaneAskari replied:
Are you sure that ```self.stale_droot``` is a good condition? I quickly implemented this : 

```python
Supervisor:
def validate(self, fgraph):
        if not hasattr(fgraph, 'has_destroyers'):
            return True
        if fgraph.has_destroyers(self.protected):
            raise gof.InconsistencyError("Trying to destroy a protected"
                                         "Variable.")
DestroyHandler:
def unpickle(self, fgraph):
        def get_destroyers_of(r):
            droot, _, root_destroyer = self.refresh_droot_impact()
            try:
                return [root_destroyer[droot[r]]]
            except Exception:
                return []
        fgraph.destroyers = get_destroyers_of

        def recursive_destroys_finder(protected_var):
            # protected_var is the idx'th input of app.
            for (app, idx) in protected_var.clients:
                if app == 'output':
                    continue
                destroy_maps = getattr(app.op, 'destroy_map', {}).values()
                # If True means that the apply node, destroys the protected_var.
                if idx in [dmap for sublist in destroy_maps for dmap in sublist]:
                    return True
                for var_idx in getattr(app.op, 'view_map', {}).keys():
                    if idx in app.op.view_map[var_idx]:
                        # We need to recursivly check the destroy_map of all the
                        # outputs that we have a view_map on.
                        if recursive_destroys_finder(app.outputs[var_idx]):
                            return True
            return False

        def has_destroyers(protected_list):
            if not self.stale_droot:
                for var in protected_list + list(fgraph.outputs):
                    if get_destroyers_of(var):
                        return True
                return False

            for protected_var in protected_list:
                if recursive_destroys_finder(protected_var):
                    return True
            return False

        fgraph.has_destroyers = has_destroyers
```

But I had many errors with the OutputGuard:
```
NotImplementedError: ('OutputGuard.0 is a view/destroyed version of more then one inputs. Currently, we only support the case where an output is a view or a destroyed version of one input.', 'Test <function <lambda> at 0x7fac379dbb18>::x2: Error occurred while computing the gradient on the following inputs: [array([[-0.07363263, -0.66992872,  0.980925  ]])]')
```

 So I guess the condition is not correct. I think the could would be nicer with your change but why do you think this will be a speedup?

nouiz replied:
I implemented it in 5e0cc2b7c89da35da27beb3302034d6befa9135d

nouiz commented on theano/compile/function_module.py:
```diff
@@ -133,6 +133,11 @@ def __init__(self, protected):
+            return True
         if not hasattr(fgraph, 'destroyers'):
             return True
         for r in self.protected + list(fgraph.outputs):
```
With the other change proposed, we could replace the full implementation of this method to do the same thing for all values of cycle_detection.

nouiz commented:
In the file compile/debugmode.py, the function _optcheck_fgraph need to be updated.

I checked this function and in fact, it should be the same as std_fgraph() + the addition of the EquivalenceTracker. It is currently different by mistake, as it wasn't updated as it should!

So, can you update this function to call std_fgraph and then add the EquivalentTracker to the fgraph. This will prevent futur error of not updating both function.

nouiz commented:
I think jenkins error isn't related to this PR and will be fixed at the next run.

ReyhaneAskari commented:
jenkins test this.

ReyhaneAskari commented:
I'm going to create another PR for this:
> In the file compile/debugmode.py, the function _optcheck_fgraph need to be updated...

ReyhaneAskari commented:
The test suit is running with ```cycle_detection = regular```. I will change it to ```fast``` when they all passed.

ReyhaneAskari commented:
jenkins test this.

ReyhaneAskari commented:
I didn't get the error for jenkins locally.

nouiz commented:
test passed. Can you redo the Theano opt profile to see if this remove all the call that are slow?

ReyhaneAskari commented:
Sure. I'll update you when the results were available.

ReyhaneAskari commented:
I added the last commit to make sure the tests will pass when the default of ```cycle_detection```  is set to ```fast```.

ReyhaneAskari commented:
I don't get jenkins' errors locally. jenkins test this.

ReyhaneAskari commented:
Here is the results of profiling. I also did the python profiling you can check them in the attached files. ```get_destroyers_of``` is no longer the appearing on the top of the list with faster. 

|               |   Function.call    | total compile time | Opt time    | Linker time   | Time since theano import |
|---------------|--------------------|--------------------|-------------|---------------|--------------------------|
| [master_2](https://github.com/Theano/Theano/files/1091779/master_2.txt)                  | 2304.337           | 469.5146           | 437.3199      | 21.41075                 | 2936.39      |
| [master_3](https://github.com/Theano/Theano/files/1091778/master_3.txt)                  | 2305.517           | 470.0153           | 437.6318      | 21.52863                 | 2937.23      |
| **avg master_2 master_3** | **2304.927**       | **469.76495**      | **437.47585** | **21.46969**             | **2936.81**  |
| [faster_2](https://github.com/Theano/Theano/files/1091776/faster_2.txt)                  | 2305.755           | 242.6038           | 211.492       | 20.39481                 | 2719.367     |
| [faster_3](https://github.com/Theano/Theano/files/1091777/faster_3.txt)                  | 2309.326           | 245.9866           | 214.8715      | 20.31912                 | 2727.227     |
| **avg faster_2 faster_3** | **2307.5405**      | **244.2952**       | **213.18175** | **20.356965**            | **2723.297** |


Memory profiling: 

||master|faster|
|----|--|--|
| peak memory cpu                      |       3.0 |       2.0 |
| peak memory gpu                      | 2293322.0 | 2186307.0 |
| peak memory cpu + gpu                | 2293324.0 | 2186309.0 |
| flag = optimizer_excluding cpu       |       3.0 |       2.0 |
| flag = optimizer_excluding gpu       | 2306834.0 | 2199204.0 |
| flag = optimizer_excluding cpu + gpu | 2306837.0 | 2199206.0 |
| allow_gc=False cpu                   |      47.0 |      47.0 |
| allow_gc=False gpu                   | 8392359.0 | 8693082.0 |
| allow_gc=False cpu + gpu             | 8392406.0 | 8693129.0 |

slefrancois commented:
jenkins test this

nouiz commented:
jenkins test this

ReyhaneAskari commented:
jenkins test this.

ReyhaneAskari commented:
jenkins test this.

nouiz commented on theano/compile/mode.py:
```diff
@@ -111,6 +111,8 @@ class AddDestroyHandler(gof.Optimizer):
 
     """
     def apply(self, fgraph):
+        if theano.config.cycle_detection == 'fast':
```
Why this change? I didn't expected this PR to change the tests that fail with cycle_detection=False. I would not be surprised that this change isn't needed and removing it "fix" the failing tests.

ReyhaneAskari replied:
This is what we had discussed previously with Pascal. When we have the ```fast```, we don't want to have the check for the ```outputguard```.  There is no failing test on this PR. All the tests have passed.

nouiz commented:
This PR skip many new test. I don't think this is good.

If the destroy handler isn't there,  then from memory, I'm in the subway,
has_destroyers won't be there, as it is the destroy handler that add it.

Le jeu. 29 juin 2017 17:55, Reyhane Askari <notifications@github.com> a
écrit :

> *@ReyhaneAskari* commented on this pull request.
> ------------------------------
>
> In theano/compile/mode.py
> <https://github.com/Theano/Theano/pull/6000#discussion_r124922287>:
>
> > @@ -111,6 +111,8 @@ class AddDestroyHandler(gof.Optimizer):
>
>      """
>      def apply(self, fgraph):
> +        if theano.config.cycle_detection == 'fast':
>
> This is what we had discussed previously with Pascal. When we have the
> fast, we don't want to have the check for the outputguard. There is no
> failing test on this PR. All the tests have passed.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6000#discussion_r124922287>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-_q1c0LnS7_lDIgcD-_uGC0ZstkNks5sJB1TgaJpZM4Ntcs->
> .
>

ReyhaneAskari commented:
Which new tests do you mean?

nouiz commented:
This PR create a decorator to mess with many tests. I don't think this is
good. We probably need to discuss this live later. This PR was just about
removing call to build the internal structure. You can't just not add the
destroy handler when cycle_detection is fast, for the simple reason that
"fast" is implemented by the destroy handler itself. So mostly, you remove
all the checks in fast, even the fast check.

On Thu, Jun 29, 2017 at 8:22 PM Reyhane Askari <notifications@github.com>
wrote:

> Which new tests do you mean?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/6000#issuecomment-312138825>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-1UjIERN8MOqDl57tEuvHLOfshutks5sJDo5gaJpZM4Ntcs->
> .
>

ReyhaneAskari commented:
After a review with @nouiz , he found that there was a minor indentation problem inside ```fast_destroy``` function (my last commit).  This change added 3 new failures which are probably due to the new check for ```destroy_handler```. I think they should be added to the ```expectedFailure_fast```.

nouiz commented:
Agreed to take the 5 new failure with expected failure in fast cyclde_detection

ReyhaneAskari commented:
Fixed the problem and rebased due to the conflicts that were introduced after the merge of the branch which removed the outputguard.

lamblin commented:
I think the issue in DebugMode is that we should skip trying to insert OutputGuard if cycle_detection is "fast". See https://github.com/Theano/Theano/pull/6044#pullrequestreview-50136155

lamblin commented:
The remaining issues seem to come from a change in the behaviour of `expectedFailure` between python 2 and 3.
For python 2, `expectedFailure` catches failures as well as Exception being raised.
For python 3, since Exceptions cause ERROR instead of FAIL, the whole thing is considered as an issue.

In that case, I think we would need to write our own decorator, consistent in python 2 and 3 instead.

ReyhaneAskari commented:
Thanks Pascal. I checked and there is no decorator in ```unittest``` to mark a test as expected error! 

So I'd suggest that we skip these tests for now:

```python

    if theano.config.cycle_detection == 'fast':
        return unittest.skip("cycle_detection is set to fast."
                             "Many inplaces are not allowed any more.")(f)
    else:
        return f
```

lamblin commented on theano/tests/unittest_tools.py:
```diff
@@ -447,11 +447,16 @@ def attempt_multiple_times(*args, **kwargs):
+                with self.assertRaises(Exception):
+                    f(*args, **kwargs)
+        test_assertion = TestAssertion()
+        return test_assertion
```
I think you can simplify this a bit by importing
```
from nose.tools import assert_raises
```
and then
```
def assertFailure_fast(f):
    if theano.config.cycle_detection == 'fast':
        def new_test(*args, **args):
            assert_raises(f, *args, **kwargs)
        return new_test
    else:
        return f
```
rather than defining and instanciating a whole new class

ReyhaneAskari replied:
Thanks, I didn't know about this one!

lamblin commented:
Also, I'm not sure why the tests for cholesky and matrix inverse with magma are failing, rather than just rejecting the optimization and failing later.

ReyhaneAskari commented:
jenkins test this.

ReyhaneAskari commented:
I did a rebase locally and the tests passed but since there were no changes I couldn't push. So I made a dummy commit. Hopefully the tests will pass now.

ReyhaneAskari commented:
I can't reproduce jenkins' errors. Do you have any ideas why this is happening? @lamblin

ReyhaneAskari commented:
I guess this is finally good to merge. Do we want the ```cycle_detection``` flag's default be ```fast``` or `regular`?
